// @statikly/funk/dist/funk.cjs.development.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);
var reactUid = require('react-uid');

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

function _objectWithoutPropertiesLoose(source, excluded) {
  if (source == null) return {};
  var target = {};
  var sourceKeys = Object.keys(source);
  var key, i;

  for (i = 0; i < sourceKeys.length; i++) {
    key = sourceKeys[i];
    if (excluded.indexOf(key) >= 0) continue;
    target[key] = source[key];
  }

  return target;
}

var Direction;

(function (Direction) {
  Direction[Direction["HORIZONTAL"] = 0] = "HORIZONTAL";
  Direction[Direction["VERTICAL"] = 1] = "VERTICAL";
})(Direction || (Direction = {}));

var TabContext = /*#__PURE__*/React.createContext({
  id: 'Please provide a unique id',
  activeTab: 0,
  numTabs: 0,
  setActiveTab: function setActiveTab() {},
  direction: Direction.HORIZONTAL
});
function useTab(index) {
  var _useContext = React.useContext(TabContext),
      activeTab = _useContext.activeTab;

  return {
    isActive: activeTab === index
  };
}
function TabGroup(_ref) {
  var _ref$initialActiveTab = _ref.initialActiveTab,
      initialActiveTab = _ref$initialActiveTab === void 0 ? 0 : _ref$initialActiveTab,
      numTabs = _ref.numTabs,
      _ref$direction = _ref.direction,
      direction = _ref$direction === void 0 ? Direction.HORIZONTAL : _ref$direction,
      children = _ref.children;
  var id = reactUid.useUID();

  var _useState = React.useState(initialActiveTab),
      activeTab = _useState[0],
      setActiveTab = _useState[1];

  var value = React.useMemo(function () {
    return {
      id: id,
      activeTab: activeTab,
      setActiveTab: setActiveTab,
      numTabs: numTabs,
      direction: direction
    };
  }, [activeTab, setActiveTab, numTabs, direction, id]);
  return React__default.createElement(TabContext.Provider, {
    value: value
  }, children);
}
function Tab(_ref2) {
  var index = _ref2.index,
      className = _ref2.className,
      activeClassName = _ref2.activeClassName,
      inactiveClassName = _ref2.inactiveClassName,
      _ref2$trigger = _ref2.trigger,
      trigger = _ref2$trigger === void 0 ? 'click' : _ref2$trigger,
      props = _objectWithoutPropertiesLoose(_ref2, ["index", "className", "activeClassName", "inactiveClassName", "trigger"]);

  var _useContext2 = React.useContext(TabContext),
      id = _useContext2.id,
      activeTab = _useContext2.activeTab,
      setActiveTab = _useContext2.setActiveTab;

  var tabRef = React.useRef();
  React.useEffect(function () {
    if (activeTab === index) {
      var _tabRef$current;

      (_tabRef$current = tabRef.current) == null ? void 0 : _tabRef$current.focus({
        preventScroll: true
      });
    }
  }, [activeTab, index]);
  var calculatedClassName = React.useMemo(function () {
    return [className, activeTab === index ? activeClassName : inactiveClassName].join(' ');
  }, [className, activeClassName, inactiveClassName, activeTab, index]);
  var interactionProps = React.useMemo(function () {
    var interactions = {};

    interactions.onClick = function () {
      return setActiveTab(index);
    };

    if (trigger === 'hover') {
      interactions.onMouseEnter = function () {
        return setActiveTab(index);
      };
    }

    return interactions;
  }, [trigger, setActiveTab, index]);
  return React__default.createElement("button", Object.assign({
    ref: tabRef,
    role: "tab",
    "aria-controls": id + "-" + index,
    "aria-selected": activeTab === index,
    tabIndex: activeTab === index ? 0 : -1,
    className: calculatedClassName
  }, interactionProps, props));
}
function TabPanel(_ref3) {
  var _ref3$as = _ref3.as,
      Component = _ref3$as === void 0 ? 'div' : _ref3$as,
      index = _ref3.index,
      className = _ref3.className,
      activeClassName = _ref3.activeClassName,
      inactiveClassName = _ref3.inactiveClassName,
      props = _objectWithoutPropertiesLoose(_ref3, ["as", "index", "className", "activeClassName", "inactiveClassName"]);

  var _useContext3 = React.useContext(TabContext),
      id = _useContext3.id,
      activeTab = _useContext3.activeTab;

  var calculatedClassName = React.useMemo(function () {
    return [className, activeTab === index ? activeClassName : inactiveClassName].join(' ');
  }, [className, activeClassName, inactiveClassName, activeTab, index]);
  return React__default.createElement(Component, Object.assign({
    style: _extends({}, props.style, {
      visibility: activeTab === index ? 'visible' : 'hidden'
    }),
    id: id + "-" + index,
    role: "tabpanel",
    className: calculatedClassName
  }, props));
}
function TabList(_ref4) {
  var _ref4$as = _ref4.as,
      Component = _ref4$as === void 0 ? 'div' : _ref4$as,
      props = _objectWithoutPropertiesLoose(_ref4, ["as"]);

  var tabListRef = React.useRef();

  var _useContext4 = React.useContext(TabContext),
      setActiveTab = _useContext4.setActiveTab,
      numTabs = _useContext4.numTabs,
      activeTab = _useContext4.activeTab,
      direction = _useContext4.direction;

  var handleKeyUp = React.useCallback(function (e) {
    var actions = {
      End: function End() {
        setActiveTab(numTabs - 1);
      },
      Home: function Home() {
        setActiveTab(0);
      }
    };

    if (direction === Direction.HORIZONTAL) {
      actions.ArrowLeft = function () {
        var previous = activeTab - 1 >= 0 ? activeTab - 1 : numTabs - 1;
        setActiveTab(previous);
      };

      actions.ArrowRight = function () {
        var next = activeTab + 1 < numTabs ? activeTab + 1 : 0;
        setActiveTab(next);
      };
    }

    if (direction === Direction.VERTICAL) {
      actions.ArrowUp = function () {
        var previous = activeTab - 1 >= 0 ? activeTab - 1 : numTabs - 1;
        setActiveTab(previous);
      };

      actions.ArrowDown = function () {
        var next = activeTab + 1 < numTabs ? activeTab + 1 : 0;
        setActiveTab(next);
      };
    }

    var action = actions[e.key];

    if (action) {
      e.preventDefault();
      action();
    }
  }, [setActiveTab, activeTab, numTabs, direction]);
  React.useEffect(function () {
    tabListRef.current.addEventListener('keyup', handleKeyUp);
    return function () {
      return tabListRef.current.removeEventListener('keyup', handleKeyUp);
    };
  }, [handleKeyUp]); // @ts-ignore

  return React__default.createElement(Component, Object.assign({
    ref: tabListRef,
    role: "tablist"
  }, props));
}
TabGroup.Tab = Tab;
TabGroup.TabPanel = TabPanel;
TabGroup.TabList = TabList;
TabGroup.useTab = useTab;
TabGroup.direction = Direction;

exports.TabGroup = TabGroup;
